### PR TITLE
Yuval generic subsystem

### PR DESCRIPTION
**Pull request by YuvalLevy1**

### CODE REVIEW

File: src/com/spikes2212/command/drivetrains/TankDrivetrain.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+ * This class represents a type of Drivetrain that its left and right sides are
+ * controlled independently, allowing it to move by giving each side a speed
+ * value separately.
+ *
+ * It can move forwards/backwards by giving both its sides an equal speed or
+ * turn by giving the sides different speeds
+ *
+ * @author Yuval Levy
+ */
+public class TankDrivetrain implements Subsystem {
```
Comment (ScarletMetal) [inline src/com/spikes2212/command/drivetrains/TankDrivetrain.java]: Make `TankDrivetrain` extend `SubsystemBase`. this is a class that implements `Subsystem` and also provides `Sendable` functionality.

File: src/com/spikes2212/command/genericsubsystem/GenericSubsystem.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+import java.util.function.Supplier;
+
+/**
+ * This class represents a generic subsystem that moves within a limitation, or
+ * without one.
+ *
+ * @author Yuval Levy
+ */
+public abstract class GenericSubsystem implements Subsystem {
```
Comment (ScarletMetal) [inline src/com/spikes2212/command/genericsubsystem/GenericSubsystem.java]: make `GenericSubsystem` extend `SubsystemBase`. see `TankDrivetrain` comment

File: src/main/java/com/spikes2212/control/JoinedPID.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.Arrays;
+import java.util.List;
+
+public class JoinedPID implements  PIDLoop{
+
+    private List<PIDLoop> pidLoops;
+
+    public JoinedPID(PIDLoop... pidLoops){
+        this.pidLoops.addAll(Arrays.asList(pidLoops));
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/JoinedPID.java]: you can remove `this` from here.

File: src/main/java/com/spikes2212/control/JoinedPID.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
+package com.spikes2212.control;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class JoinedPID implements  PIDLoop{
+
+    private List<PIDLoop> pidLoops;
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/JoinedPID.java]: you need to instantiate `pidLoops` either here or in the constructor.

File: src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithPID.java [hunk 5]
```diff
@@ -0,0 +1,4 @@
+package com.spikes2212.command.drivetrains.commands;
+
+public class DriveArcadeWithPID {
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/command/drivetrains/commands/DriveArcadeWithPID.java]: you may remove this class.

File: src/main/java/com/spikes2212/command/genericsubsystem/commands/RiseGenericSubsystem.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ *
+ * This command rise the speed of a {@link GenericSubsystem} linearly so it will
+ * reach a wanted speed after a given time.
+ *
+ * @author Yuval Levy
+ *
+ */
+public class RiseGenericSubsystem extends MoveGenericSubsystem {
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/command/genericsubsystem/commands/RiseGenericSubsystem.java]: we can rename this command to `AccelerateSubsystem` or `AccelerateGenericSubsystem`. this command should just make a subsystem reach a sertain speed. we can then hook it up with other command using the `andThen` method from command.

File: src/main/java/com/spikes2212/control/JoinedPID.java [hunk 7]
```diff
@@ -0,0 +1,43 @@
+package com.spikes2212.control;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class JoinedPID implements PIDLoop {
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/JoinedPID.java]: it is better to name this class `JoinedPIDLoop` or something. just so we stick to the naming convention we created for all the other implementations of `PIDLoop`.

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 8]
```diff
@@ -117,14 +165,22 @@ public void disable() {
     }
 
 
-    public void enableContinuousInput(double minimumValue, double maximumValue) {
+    private void enableContinuousInput(double minimumValue, double maximumValue) {
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: there is no need for this one-liner function. you can just run it in the constructor.

File: src/main/java/com/spikes2212/control/RioPIDLoop.java [hunk 9]
```diff
@@ -117,14 +165,22 @@ public void disable() {
...
         this.controller.enableContinuousInput(minimumValue, maximumValue);
     }
 
-    public void disableContinuousInput() {
+    private void disableContinuousInput() {
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/control/RioPIDLoop.java]: this function is private and never used. so we can just remove it.

File: src/main/java/com/spikes2212/control/JoinedPID.java (hunk 3 above)
Comment (YuvalLevy1) [reply src/main/java/com/spikes2212/control/JoinedPID.java]: doing so would make pidLoops to be the given variable and not the class' variable

File: src/main/java/com/spikes2212/command/genericsubsystem/commands/AccelerateGenericSubsystem.java [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+import edu.wpi.first.wpilibj.Timer;
+
+/**
+ * This command accelerate the speed of a {@link GenericSubsystem} linearly so it will
+ * reach a wanted speed after a given time.
+ *
+ * @author Yuval Levy
+ */
+public class AccelerateGenericSubsystem extends MoveGenericSubsystem {
```
Comment (ScarletMetal) [inline src/main/java/com/spikes2212/command/genericsubsystem/commands/AccelerateGenericSubsystem.java]: please make this command to just accelerate the subsystem, and finish when it is at the needed speed. 
We can use command groups, and specifically the `andThen` method in command in order to make the subsystem stay at the desired speed afterwards.

**Pull request merged**